### ISSUE TITLE
F821 (E0601) False negative

### ISSUE THREAD

**Issue by pa5555** [astral-sh/ruff#16839]
### Summary

Both pylint and ruff fails to detect mistake in the code, see attached link

https://play.ruff.rs/f229c153-a059-4efe-99d5-2a600f0ecc70

### Version

0.11.0

**Comment by ntBre** [comment]
Thanks for the report! There are some other open issues around F821 (https://github.com/astral-sh/ruff/issues/6242 and https://github.com/astral-sh/ruff/issues/9858, for example), but those seem to be related to `del`. I think it's worth tracking this separately.

I think we would need to know that the assignment to `x` can raise an exception, so this probably requires better type inference. Or we could take the approach described in the [pylint docs](https://pylint.readthedocs.io/en/latest/user_guide/messages/error/used-before-assignment.html):

> Assignments in try blocks are assumed not to have occurred when evaluating associated except/finally blocks.

(pasting the playground code for easier quick reference)

```python
# ruff says all good
# pylint says: E0601: Using variable 'x' before assignment (used-before-assignment)
try:
    x = 1 / 0
    #if x:
    #    ...
except Exception:
    print(x)


# ruff says all good
# pylint says all good
try:
    x = 1 / 0
    if x:
        ...
except Exception:
    print(x)
```

**Comment by MichaReiser** [comment]
I think the most correct would be to mark every variable defined inside a `try` to be possibly unbound in the `except` handlers and any code coming after. However, I worry that this is too strict for most users and should probably be its own rule. At least, this is what we do in our new type checker ([playground](https://playknot.ruff.rs/2ac1f0fb-c2b6-477a-bea5-3a6a77edceb7))

**Comment by afoglia** [comment]
A summary of what was in [#23877](https://github.com/astral-sh/ruff/issues/23877)

Given this code
```python
class Foo:
    def __init__(self):
        # init stuff that might throw

    def handle_exception(self):
        pass

try:
    foo = Foo()
    # other stuff that might throw
except Exception:
    foo.handle_exception()
    x = 10

print(x)
```

`ruff check` does not report:
1. That foo in the except clause may be undefined
2. That the x in the print x may be undefined

Pylint's E0601 rule does trigger on both of those lines. PyCharm's linter does report a warning at 1. (I don't run PyCharm myself, so I haven't checked if it reports on line 2.)

I think both of these are useful checks.

Given there is some concern that it may return false positives, or be too strict, doing this as part of a new PLE0601 seems to make the most sense. It's not currently a rule, so users would have to treat it like any other new rule. And it matches pylint behavior, not pyflakes.
